core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala

```diff
@@ -463,31 +469,49 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
       }
     }
 
+    def determineInvites(emailsNotFound: Seq[WorkspaceACLUpdate], inviteUsersNotFound: Boolean): Future[Seq[WorkspaceACLUpdate]] = {
+      dataSource.inTransaction { dataAccess =>
+        withWorkspaceContext(workspaceName, dataAccess) { workspaceContext =>
+          dataAccess.workspaceQuery.getInvites(workspaceContext.workspaceId) flatMap { invites =>
+            val (toDelete, toSave) = emailsNotFound.partition(_.accessLevel.equals(WorkspaceAccessLevels.NoAccess))
+            val dedupedInvites = toSave.filterNot(update => invites.contains((update.email, update.accessLevel)))
+
+            if(inviteUsersNotFound) {
+              DBIO.sequence(dedupedInvites.map { invite =>
+                dataAccess.workspaceQuery.saveInvite(workspaceContext.workspaceId, userInfo.userSubjectId, WorkspaceACLUpdate(invite.email, invite.accessLevel))
+              }) flatMap { _ =>
+                DBIO.sequence(toDelete.map { invite =>
+                  dataAccess.workspaceQuery.removeInvite(workspaceContext.workspaceId, invite.email)
+                }) map { _ => (dedupedInvites ++ toDelete) }
+              }
+            } else DBIO.successful(dedupedInvites ++ toDelete) //else only return invites that don't already exist
+          }
+        }
+      }
+    }
+
+    def getUsersUpdatedResponse(actualChangesToMake: Map[Either[RawlsUserRef, RawlsGroupRef], WorkspaceAccessLevel]): Seq[WorkspaceACLUpdateResponse] = {
+      actualChangesToMake.map { case (member, accessLevel) =>
+        member match {
+          case Left(userRef) => WorkspaceACLUpdateResponse(userRef.userSubjectId.value, accessLevel)
+          case Right(groupRef) => WorkspaceACLUpdateResponse(groupRef.groupName.value, accessLevel)
+        }
+      }.toSeq
+    }
+
     val userServiceRef = context.actorOf(UserService.props(userServiceConstructor, userInfo))
     for {
```

Comment:
I think this for comprehension needs a little work. The steps should be

1. overwriteGroupMessagesFuture (no change)
1. Future.traverse(overwriteGroupMessages) (no change)
1. remove appropriate invites
1. maybe invite users which returns the actual list of invites sent

All the while progressively filtering emailsNotFound (i.e. removing invites also removes some unknown emails and inviting will remove any outstanding emails and maybe all the rest). At the end you have a list of emails not found and invites sent and you don't need to do a further check of inviteUsersNotFound.